<a href="https://colab.research.google.com/github/SakibAmin/CSC-448/blob/main/Sakib_Amin_CSC_448_Assignment_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Imports section

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Lasso

## Part 1. Loading the dataset

In [ ]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data

url = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
df = pd.read_csv(url)

df.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)


X = df.iloc[:, :-1].values
Y = df.iloc[:, -1].values
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.1)

#print(X)
#print(Y)
#print(X_train)
#print (X_test)
#print (y_train)
#print (y_test)


## Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set
# Create a sample datapoint and predict the output of that sample with the trained model
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

line = LinearRegression()
line.fit(X_train, y_train)

sample = np.array([[324, 587]])
print("Predicted Output = ", line.predict(sample))

score = line.score(X_test, y_test)
print("Score = ", score)

coefficents = line.coef_
intercept = line.intercept_
print("Coefficients = ", coefficents)
print("Intercept: ", intercept)

Predicted Output =  [472920.8860303]
Score =  0.8581793565395621
Coefficients =  [ 872.04797529 1045.88259667]
Intercept:  -423555.74220925337


The score was 0.85 or 85% meaning that the model can predict with 85% accuracy. Using the Coeffiecients and the intercept we can create make the following $h(x)$ function

$ h(x) = (872.04a) + (1045.88b) - 423555.74 $

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# Report on their finding and their significance

c_score = cross_val_score(line, X, Y)
print(c_score)

[0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]


The cross_val_score function allows you to determine what the potential accuaracy of your model could be if you were to run the model multiple times. In this case the cross_val_score function is giving a range between 83% to 87% meaning that generally our model will perform with that range of accuracy

## Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Report on the metrics and output the resultant equation as you did in Part 3.

poly = PolynomialFeatures(degree=2)
poly_X = poly.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(poly_X, Y, test_size=0.1)

model = Lasso()
model.fit(X_train, y_train)

score = model.score(X_test, y_test)
print("Score =", score)
coefficents = model.coef_
intercept = model.intercept_
print("Coefficients = ", coefficents)
print("Intercept = ", intercept)


Score = 0.9999999999910508
Coefficients =  [ 0.00000000e+00  1.19840878e+01 -6.48933404e-03  1.32175929e-05
  2.00000355e+00  2.85764689e-02]
Intercept =  4.198143994843122


If we look at our score value we can see that the model has a 99% accuracy meaning that it has a very high and accurate perdictive model. Therefore we can say that the data may be closer to a polynomial trend rather than a linear trend. If we were to create a $h(x)$ function for the followinng model it would look like the following

$ h(x) = ((1.32175929e-05)a^2) + ((2.85764689e-02)b^2) + (2.00000355ab) + ((1.19840878e+01)a) + ((-6.48933404e-03)b) + 4.198143994843122 $